### ISSUE TITLE
KbestHaplotypeFinder.removeCyclesAndVerticesThatDontLeadToSinks() is recursive and inefficient. 

### ISSUE THREAD

**Issue by jamesemery** [broadinstitute/gatk#5954]
In the process of working to generify the `KBestHaplotypeFinder` it has come to my attention that the method `KbestHaplotypeFinder.removeCyclesAndVerticesThatDontLeadToSinks()` seems inefficient. Specifically it calls out to `findGuiltyVerticesAndEdgesToRemoveCycles()` which is a recursive method that attempts to search every child path through the graph and determine which ones either don't lead to sinks or loop back on themselves, adding rule breaking edges to a list. The performance cost seems like it would relate to the size of the graph which would have a bigger impact on the non-seq graph generated after #5922.

Currently no information is preserved between edges meaning that much work will be repeated (for instance a dynamic approach could save the work of revisiting already visited edges). This may be complicated somewhat by the pruning mechanism this method uses to generate parent vertexes. 

This is low priority since this method will likely be entirely replaced by a new approach in #5923 later anyway.
